File: clients/typescript/src/conversation/secret-registry.ts
/**
 * Secret Registry for Conversations
 *
 * Manages secrets and injects them into commands when needed.
 * Provides secret masking in output to prevent accidental exposure.
 * This mirrors the Python SDK's SecretRegistry class.
 */

import { SecretValue } from '../types/base';

// Re-export so existing consumers of this module still find the type
export type { SecretValue } from '../types/base';

/**
 * Secret source types
 */
export type SecretSourceKind = 'static' | 'callable';

/**
 * Base interface for secret sources
 */
export interface SecretSource {
  kind: SecretSourceKind;
  getValue(): string | null;
}

/**
 * Static secret source - stores a fixed value
 */
export class StaticSecretSource implements SecretSource {
  readonly kind: SecretSourceKind = 'static';
  private value: string;

  constructor(value: string) {
    this.value = value;
  }

  getValue(): string {
    return this.value;
  }
}

/**
 * Callable secret source - evaluates a function to get the value
 */
export class CallableSecretSource implements SecretSource {
  readonly kind: SecretSourceKind = 'callable';
  private getter: () => string;

  constructor(getter: () => string) {
    this.getter = getter;
  }

  getValue(): string | null {
    try {
      return this.getter();
    } catch {
      return null;
    }
  }
}

/**
 * Wrap a SecretValue into a SecretSource
 */
function wrapSecret(value: SecretValue): SecretSource {
  if (typeof value === 'function') {
    return new CallableSecretSource(value);
  }
  return new StaticSecretSource(value);
}

/**
 * Manages secrets and injects them into bash commands when needed.
 *
 * The secret registry stores a mapping of secret keys to SecretSources
 * that retrieve the actual secret values. When a bash command is about to be
 * executed, it scans the command for any secret keys and provides the
 * corresponding environment variables.
 *
 * Additionally, it tracks the latest exported values to enable consistent masking
 * even when callable secrets fail on subsequent calls.
 *
 * Example:
 * ```typescript
 * const registry = new SecretRegistry();
 * registry.updateSecrets({
 *   'API_KEY': 'sk-secret-key',
 *   'DYNAMIC_TOKEN': () => getToken(),
 * });
 *
 * // Get env vars for a command
 * const envVars = registry.getSecretsAsEnvVars('curl -H "Authorization: Bearer $API_KEY" ...');
 *
 * // Mask secrets in output
 * const maskedOutput = registry.maskSecretsInOutput('Response with sk-secret-key visible');
 * // Returns: 'Response with <secret-hidden> visible'
 * ```
 */
export class SecretRegistry {
  /** Map of secret key to secret source */
  private secretSources: Map<string, SecretSource> = new Map();

  /** Cache of successfully exported values for masking */
  private exportedValues: Map<string, string> = new Map();

  /**
   * Add or update secrets in the registry.
   *
   * @param secrets - Dictionary mapping secret keys to values or callable functions
   */
  updateSecrets(secrets: Record<string, SecretValue>): void {
    for (const [key, value] of Object.entries(secrets)) {
      this.secretSources.set(key, wrapSecret(value));
    }
  }

  /**
   * Remove a secret from the registry.
   *
   * @param key - The secret key to remove
   */
  removeSecret(key: string): void {
    this.secretSources.delete(key);
    this.exportedValues.delete(key);
  }

  /**
   * Clear all secrets from the registry.
   */
  clearSecrets(): void {
    this.secretSources.clear();
    this.exportedValues.clear();
  }

  /**
   * Get the number of registered secrets.
   */
  get size(): number {
    return this.secretSources.size;
  }

  /**
   * Get all registered secret keys.
   */
  get keys(): string[] {
    return Array.from(this.secretSources.keys());
  }

  /**
   * Find all secret keys mentioned in the given text.
   * Uses case-insensitive matching.
   *
   * @param text - The text to search for secret keys
   * @returns Set of secret keys found in the text
   */
  findSecretsInText(text: string): Set<string> {
    const foundKeys = new Set<string>();
    const lowerText = text.toLowerCase();

    for (const key of this.secretSources.keys()) {
      if (lowerText.includes(key.toLowerCase())) {
        foundKeys.add(key);
      }
    }

    return foundKeys;
  }

  /**
   * Get secrets that should be exported as environment variables for a command.
   *
   * @param command - The bash command to check for secret references
   * @returns Dictionary of environment variables to export (key -> value)
   */
  getSecretsAsEnvVars(command: string): Record<string, string> {
    const foundSecrets = this.findSecretsInText(command);

    if (foundSecrets.size === 0) {
      return {};
    }

    const envVars: Record<string, string> = {};

    for (const key of foundSecrets) {
      const source = this.secretSources.get(key);
      if (!source) continue;

      const value = source.getValue();
      if (value) {
        envVars[key] = value;
        this.exportedValues.set(key, value);
      }
    }

    return envVars;
  }

  /**
   * Mask secret values in the given text.
   *
   * This method uses the currently exported values to ensure comprehensive masking.
   * It replaces all known secret values with '<secret-hidden>'.
   *
   * @param text - The text to mask secrets in
   * @returns Text with secret values replaced by <secret-hidden>
   */
  maskSecretsInOutput(text: string): string {
    if (!text) {
      return text;
    }

    let maskedText = text;

    // Mask using currently exported values
    for (const value of this.exportedValues.values()) {
      if (value) {
        // Use a global replace
        maskedText = maskedText.split(value).join('<secret-hidden>');
      }
    }

    // Also try to get fresh values from sources (in case they weren't exported yet)
    for (const [key, source] of this.secretSources) {
      try {
        const value = source.getValue();
        if (value && !this.exportedValues.has(key)) {
          maskedText = maskedText.split(value).join('<secret-hidden>');
        }
      } catch {
        // Ignore errors when trying to mask - we'll use cached values
      }
    }

    return maskedText;
  }

  /**
   * Check if a specific secret key is registered.
   *
   * @param key - The secret key to check
   * @returns True if the key is registered
   */
  hasSecret(key: string): boolean {
    return this.secretSources.has(key);
  }

  /**
   * Serialize the registry state for persistence.
   * Note: Callable secrets cannot be serialized and will be omitted.
   * Secret values are redacted by default for security.
   *
   * @param exposeSecrets - If true, include actual secret values (use with caution!)
   * @returns Serialized registry state
   */
  serialize(exposeSecrets: boolean = false): Record<string, string | null> {
    const result: Record<string, string | null> = {};

    for (const [key, source] of this.secretSources) {
      if (source.kind === 'callable') {
        // Callable secrets can't be serialized
        result[key] = null;
      } else if (exposeSecrets) {
        result[key] = source.getValue();
      } else {
        result[key] = '**********';
      }
    }

    return result;
  }

  /**
   * Restore secrets from serialized state.
   * Only static secrets can be restored.
   *
   * @param state - Previously serialized state
   */
  deserialize(state: Record<string, string | null>): void {
    for (const [key, value] of Object.entries(state)) {
      if (value !== null && value !== '**********') {
        this.secretSources.set(key, new StaticSecretSource(value));
      }
    }
  }
}

File: openhands-sdk/openhands/sdk/conversation/secret_registry.py
"""Secrets manager for handling sensitive data in conversations."""

import re
import time
from collections.abc import Callable, Collection, Mapping
from enum import Enum
from threading import RLock
from typing import Any, Final

from pydantic import BaseModel, Field, PrivateAttr, SecretStr

from openhands.sdk.logger import get_logger
from openhands.sdk.secret import SecretSource, SecretValue, StaticSecret
from openhands.sdk.utils.models import OpenHandsModel


logger = get_logger(__name__)

# Back-off before retrying a failed source; a failed lookup masks nothing anyway.
FAILED_LOOKUP_RETRY_SECONDS: Final[float] = 60.0


def _mask_value(value: Any, mask: Callable[[str], str]) -> Any:
    """Recursively mask every ``str`` reachable from ``value``."""
    match value:
        case Enum():
            # A str-subclass enum is a str, but masking it would downgrade the
            # member to a plain str and break the field's serialization. Its
            # vocabulary is fixed, so it can never hold a secret anyway.
            return value
        case str():
            return mask(value)
        case BaseModel():
            return _mask_model(value, mask)
        case list():
            return [_mask_value(item, mask) for item in value]
        case tuple():
            items = [_mask_value(item, mask) for item in value]
            # Rebuild a NamedTuple through its own constructor; tuple(items)
            # would downgrade it the same way masking an Enum member does.
            return type(value)(*items) if hasattr(value, "_fields") else tuple(items)
        case set() | frozenset():
            return type(value)(_mask_value(item, mask) for item in value)
        case dict():
            return {key: _mask_value(item, mask) for key, item in value.items()}
        case _:
            return value


def _mask_model[ModelT: BaseModel](model: ModelT, mask: Callable[[str], str]) -> ModelT:
    """Rebuild ``model`` with every nested string masked.

    ``model_copy`` is used rather than a validate round-trip so private
    attributes survive and no field is re-coerced.
    """
    updates = {
        name: _mask_value(getattr(model, name), mask)
        for name in type(model).model_fields
    }
    return model.model_copy(update=updates)


class StreamOutputMask:
    """Masks text that arrives in pieces.

    A per-chunk masker cannot see a secret split across chunk boundaries, so
    this holds back the shortest suffix that could still grow into one: the
    longest registered value minus one character.
    """

    def __init__(self, pattern: re.Pattern[str] | None, max_len: int) -> None:
        self._pattern = pattern
        self._max_len = max_len
        self._held = ""

    def feed(self, text: str) -> str:
        """Return the masked text that is safe to release now."""
        pattern = self._pattern
        if pattern is None:
            return text
        self._held += text
        return self._release(pattern, max(len(self._held) - self._max_len + 1, 0))

    def flush(self) -> str:
        """Return what is still held back; no more input is coming."""
        pattern = self._pattern
        if pattern is None:
            return ""
        return self._release(pattern, len(self._held))

    def _release(self, pattern: re.Pattern[str], cut: int) -> str:
        out: list[str] = []
        pos = 0
        end = cut
        for match in pattern.finditer(self._held):
            # A match starting before the cut is already maximal: the cut
            # leaves max_len-1 characters of lookahead behind it.
            if match.start() >= cut:
                break
            out.append(self._held[pos : match.start()])
            out.append("<secret-hidden>")
            pos = match.end()
            end = max(end, match.end())
        out.append(self._held[pos:end])
        self._held = self._held[end:]
        return "".join(out)


class SecretRegistry(OpenHandsModel):
    """Manages secrets and injects them into bash commands when needed.

    The secret registry stores a mapping of secret keys to SecretSources
    that retrieve the actual secret values. When a bash command is about to be
    executed, it scans the command for any secret keys and injects the corresponding
    environment variables.

    Secret sources will redact / encrypt their sensitive values as appropriate when
    serializing, depending on the content of the context. If a context is present
    and contains a 'cipher' object, this is used for encryption. If it contains a
    boolean 'expose_secrets' flag set to True, secrets are dunped in plain text.
    Otherwise secrets are redacted.

    Additionally, it tracks the latest exported values to enable consistent masking
    even when callable secrets fail on subsequent calls.
    """

    secret_sources: dict[str, SecretSource] = Field(default_factory=dict)
    _exported_values: dict[str, str] = PrivateAttr(default_factory=dict)
    _exported_values_lock: RLock = PrivateAttr(default_factory=RLock)
    _failed_lookups: dict[str, float] = PrivateAttr(default_factory=dict)

    def track_exported_values(self, values: Mapping[str, str]) -> None:
        """Track values for output masking."""
        with self._exported_values_lock:
            self._exported_values.update({k: v for k, v in values.items() if v})

    def update_secrets(
        self,
        secrets: Mapping[str, SecretValue],
    ) -> None:
        """Add or update secrets in the manager.

        Args:
            secrets: Dictionary mapping secret keys to either string values
                    or callable functions that return string values
        """
        secret_sources = {name: _wrap_secret(value) for name, value in secrets.items()}
        self.secret_sources.update(secret_sources)

    def find_secrets_in_text(self, text: str) -> set[str]:
        """Find all secret keys mentioned in the given text.

        Args:
            text: The text to search for secret keys

        Returns:
            Set of secret keys found in the text
        """
        found_keys = set()
        for key in self.secret_sources.keys():
            if key.lower() in text.lower():
                found_keys.add(key)
        return found_keys

    def get_secrets_as_env_vars(self, command: str) -> dict[str, str]:
        """Get secrets that should be exported as environment variables for a command.

        Args:
            command: The bash command to check for secret references

        Returns:
            Dictionary of environment variables to export (key -> value)
        """
        found_secrets = self.find_secrets_in_text(command)

        if not found_secrets:
            return {}

        logger.debug(f"Found secrets in command: {found_secrets}")

        env_vars = {}
        for key in found_secrets:
            try:
                source = self.secret_sources[key]
                value = source.get_value()
                if value:
                    env_vars[key] = value
                    self.track_exported_values({key: value})
            except Exception as e:
                logger.error(f"Failed to retrieve secret for key '{key}': {e}")
                continue

        logger.debug(f"Prepared {len(env_vars)} secrets as environment variables")
        return env_vars

    def get_all_secrets_as_env_vars(
        self, exclude: Collection[str] | None = None
    ) -> dict[str, str]:
        """Resolve every registered secret to an env-var mapping.

        Unlike :meth:`get_secrets_as_env_vars`, which name-scans a single
        command and injects only the secrets it references, this resolves the
        whole registry. It is for opaque consumers (e.g. an ACP CLI subprocess)
        that cannot be name-scanned per command and must receive their
        credentials upfront. Resolved values are tracked for output masking, and
        lookup failures are skipped rather than raised.

        Note: this injects the *whole* registry; least-privilege scoping
        (provider creds + an explicit allowlist only) is deferred to #1039
        task 6.

        Args:
            exclude: Secret names to skip — e.g. keys a higher-precedence tier
                will set anyway, or file-content secrets materialised to disk
                (avoids a wasted, possibly remote, ``get_value()``).

        Returns:
            Dictionary of environment variables to export (key -> value),
            omitting empty values and excluded names.
        """
        skip = set(exclude or ())
        env_vars: dict[str, str] = {}
        for name in self.secret_sources:
            if name in skip:
                continue
            value = self.get_secret_value(name)
            if value:
                env_vars[name] = value
        return env_vars

    def mask_secrets_in_output(self, text: str) -> str:
        """Mask secret values in the given text.

        Masks the last resolved value of every registered secret, not only the
        exported ones: a value can reach the output without the command ever
        referencing its name (e.g. a token in a git remote URL). Cached values
        are not re-resolved, so a rotated secret masks its previous value.

        Args:
            text: The text to mask secrets in

        Returns:
            Text with secret values replaced by <secret-hidden>
        """
        if not text:
            return text

        # Resolve uncached sources, backing off on failure: get_value() may do
        # blocking network I/O and masking runs per output and per ACP chunk.
        now = time.monotonic()
        for key in list(self.secret_sources):
            if key in self._exported_values:
                continue
            failed_at = self._failed_lookups.get(key)
            if failed_at is not None and now - failed_at < FAILED_LOOKUP_RETRY_SECONDS:
                continue
            if not self.get_secret_value(key):
                self._failed_lookups[key] = now

        masked_text = text
        with self._exported_values_lock:
            exported_values = tuple(self._exported_values.values())
        for value in exported_values:
            if value:
                masked_text = masked_text.replace(value, "<secret-hidden>")

        return masked_text

    def compile_stream_mask(self) -> StreamOutputMask:
        """Return a streaming masker over the values resolved *so far*.

        For hot paths that cannot afford :meth:`mask_secrets_in_output`, which
        resolves uncached sources first — ``get_value()`` may block on network
        I/O, and on the sync agent path that runs under the state lock. Masks
        less in exchange: use it for progress, not for the durable record.
        """
        with self._exported_values_lock:
            values = {value for value in self._exported_values.values() if value}
        if not values:
            return StreamOutputMask(None, 0)
        # Longest first so an overlapping shorter value cannot mask half of a
        # longer one and leave the rest in cleartext.
        pattern = re.compile(
            "|".join(re.escape(v) for v in sorted(values, key=len, reverse=True))
        )
        return StreamOutputMask(pattern, max(len(v) for v in values))

    def mask_secrets_in_model[ModelT: BaseModel](self, model: ModelT) -> ModelT:
        """Return ``model`` with secret values masked in every nested string.

        Masking one known text field is not enough for tool output:
        ``Observation.to_llm_content`` is overridable, and several tools build
        what the model sees out of their own fields rather than ``content``.
        Walking the whole model is what keeps tool #14 covered by default.
        """
        return _mask_model(model, self.mask_secrets_in_output)

    def get_secret_infos(self) -> list[dict[str, str | None]]:
        """Get secret information (name and description) for prompt inclusion.

        Returns:
            List of dictionaries with 'name' and 'description' keys.
            Returns an empty list if no secrets are registered.
            Description will be None if not available.
        """
        if not self.secret_sources:
            return []
        secret_infos = []
        for name, source in self.secret_sources.items():
            description = source.description
            secret_infos.append({"name": name, "description": description})
        return secret_infos

    def get_secret_value(self, name: str) -> str | None:
        """Look up a single secret value by name.

        This method retrieves the value of a specific secret. It's designed
        to be passed as a callback to functions that need secret lookup
        (e.g., expand_mcp_variables) without exposing all secrets at once.

        Retrieved values are tracked in _exported_values for consistent masking
        in command outputs.

        Args:
            name: The name of the secret to retrieve.

        Returns:
            The secret value if found and successfully retrieved, None otherwise.

        Note:
            Returns None for both missing secrets and retrieval failures.
            Retrieval errors (network, auth, etc.) are logged as warnings.
        """
        source = self.secret_sources.get(name)
        if source is None:
            return None
        try:
            value = source.get_value()
            if value:
                self.track_exported_values({name: value})
            return value
        except (OSError, TimeoutError) as e:
            # Network/IO errors - likely transient, log and return None
            logger.warning(
                f"Transient error retrieving secret '{name}' "
                f"(may retry later): {type(e).__name__}: {e}"
            )
            return None
        except (ValueError, KeyError, TypeError) as e:
            # Configuration/data errors - likely permanent
            logger.warning(
                f"Configuration error for secret '{name}': {type(e).__name__}: {e}"
            )
            return None
        except Exception as e:
            # Unexpected errors - log with full details for debugging
            logger.warning(
                f"Unexpected error retrieving secret '{name}': {type(e).__name__}: {e}"
            )
            return None


def _wrap_secret(value: SecretValue) -> SecretSource:
    """Convert the value given to a secret source"""
    if isinstance(value, SecretSource):
        return value
    if isinstance(value, str):
        return StaticSecret(value=SecretStr(value))
    raise ValueError("Invalid SecretValue")
